File: src/main/java/std/nooook/readinggardenkotlin/common/config/WebMvcConfig.java
package std.nooook.readinggardenkotlin.common.config;

import java.nio.file.Path;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.CacheControl;
import org.springframework.web.servlet.config.annotation.ResourceHandlerRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@Configuration(proxyBeanMethods = false)
public class WebMvcConfig implements WebMvcConfigurer {

    private final String imagesRoot;

    public WebMvcConfig(@Value("${app.storage.images-root:/opt/reading-garden/data/images}") String imagesRoot) {
        this.imagesRoot = imagesRoot;
    }

    @Override
    public void addResourceHandlers(ResourceHandlerRegistry registry) {
        String normalizedImagesRoot = Path.of(this.imagesRoot).toAbsolutePath().normalize().toString();
        registry
                .addResourceHandler("/images/**")
                .addResourceLocations("file:" + normalizedImagesRoot + "/")
                .setCacheControl(CacheControl.noCache());
    }
}

File: src/main/kotlin/std/nooook/readinggardenkotlin/common/config/WebMvcConfig.kt
package std.nooook.readinggardenkotlin.common.config

import jakarta.servlet.MultipartConfigElement
import java.nio.file.Path
import org.springframework.boot.servlet.MultipartConfigFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.CacheControl
import org.springframework.util.unit.DataSize
import org.springframework.web.servlet.config.annotation.ResourceHandlerRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
import std.nooook.readinggardenkotlin.common.storage.StorageProperties

@Configuration
class WebMvcConfig(
    private val storageProperties: StorageProperties,
    @Value("\${spring.servlet.multipart.max-file-size:10MB}")
    private val maxFileSize: DataSize,
    @Value("\${spring.servlet.multipart.max-request-size:10MB}")
    private val maxRequestSize: DataSize,
) : WebMvcConfigurer {
    @Bean
    fun multipartConfigElement(): MultipartConfigElement {
        val factory = MultipartConfigFactory()
        factory.setLocation(Path.of(storageProperties.imagesRoot, "multipart-temp").toString())
        factory.setFileSizeThreshold(DataSize.ofBytes(0))
        factory.setMaxFileSize(maxFileSize)
        factory.setMaxRequestSize(maxRequestSize)
        return factory.createMultipartConfig()
    }

    override fun addResourceHandlers(registry: ResourceHandlerRegistry) {
        val imagesRoot = Path.of(storageProperties.imagesRoot).toAbsolutePath().normalize().toString()
        registry
            .addResourceHandler("/images/**")
            .addResourceLocations("file:$imagesRoot/")
            .setCacheControl(CacheControl.noCache())
    }
}
